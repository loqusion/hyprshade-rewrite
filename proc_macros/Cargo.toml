[package]
name = "proc_macros"
version = "0.1.0"
publish = false
edition = "2021"

[dependencies]
lazy_static = "1.5.0"
quote = "1.0.36"
syn = "2.0.74"

[lib]
proc-macro = true
