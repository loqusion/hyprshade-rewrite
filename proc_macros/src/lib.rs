use std::env;

use lazy_static::lazy_static;
use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input, ItemFn,
};

lazy_static! {
    static ref IS_HYPRLAND_RUNNING: bool = env::var_os("HYPRLAND_INSTANCE_SIGNATURE").is_some();
}

/// Attribute macro declaring a test function that will be ignored if the test is not running in a
/// Hyprland instance.
///
/// In other words, a test not running in a Hyprland instance should be roughly equivalent to the
/// following:
///
/// ```no_run
/// #[test]
/// #[ignore = "requires hyprland"]
/// fn test() { /* ... */ }
/// ```
///
/// **WARNING:** Once tests have been run, the decision to ignore tests will not be re-evaluated
/// until recompilation is triggered, e.g. by changing a file. This can lead to tests starting to
/// fail if the environment changes between runs.
#[proc_macro_attribute]
pub fn hyprland_test(attr: TokenStream, item: TokenStream) -> TokenStream {
    let item = parse_macro_input!(item as ItemFn);
    // Verify there are no attribute arguments
    let _ = parse_macro_input!(attr as HyprlandTestAttribute);

    let ignore_attr = if *IS_HYPRLAND_RUNNING {
        None
    } else {
        Some(quote! {
            #[ignore = "requires hyprland"]
        })
    };

    let expanded = quote! {
        #[test]
        #ignore_attr
        #item
    };

    TokenStream::from(expanded)
}

#[derive(Debug, Clone)]
struct HyprlandTestAttribute {}

impl Parse for HyprlandTestAttribute {
    fn parse(_input: ParseStream) -> syn::Result<Self> {
        Ok(Self {})
    }
}
