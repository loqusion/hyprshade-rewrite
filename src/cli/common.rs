use std::{collections::HashMap, convert::Infallible, str::FromStr};

pub const SHADER_HELP: &str = "Which shader to turn on";
pub const SHADER_HELP_LONG: &str = "\
    Which shader to turn on\n\
\n\
    May be a name (e.g. `blue-light-filter`)\n\
    or a path (e.g. `~/.config/hypr/shaders/blue-light-filter.glsl`)\
";

#[derive(Debug, Clone)]
pub struct ArgVar(String);

impl FromStr for ArgVar {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(Self(s.to_string()))
    }
}

impl TryFrom<ArgVar> for HashMap<String, mustache::Data> {
    type Error = ArgVarError;

    fn try_from(value: ArgVar) -> Result<Self, Self::Error> {
        let ArgVar(s) = value;
        match s.split('=').collect::<Vec<_>>()[..] {
            ["", _] => Err(ArgVarError::EmptyKey(s)),
            [_, ""] => Err(ArgVarError::EmptyValue(s)),
            [dotted_key, value] => {
                let mut key_split = dotted_key.split('.').rev();
                let mut map: HashMap<String, mustache::Data> = HashMap::from([(
                    key_split.next().unwrap().to_string(),
                    mustache::Data::String(value.to_string()),
                )]);

                for key in key_split {
                    map = HashMap::from([(key.to_string(), mustache::Data::Map(map))]);
                }
                Ok(map)
            }
            [_, _, ..] => Err(ArgVarError::TooManyEquals(s)),
            [_] => Err(ArgVarError::NoEquals(s)),
            [] => Err(ArgVarError::NoContent),
        }
    }
}

impl FromIterator<ArgVar> for Result<mustache::Data, Vec<ArgVarError>> {
    fn from_iter<T: IntoIterator<Item = ArgVar>>(iter: T) -> Self {
        let mut map: HashMap<String, mustache::Data> = HashMap::new();

        let (things, errors): (Vec<_>, Vec<_>) = iter
            .into_iter()
            .map(|arg_var| arg_var.try_into())
            .partition(Result::is_ok);
        let things: Vec<_> = things.into_iter().map(Result::unwrap).collect();
        let mut errors: Vec<_> = errors.into_iter().map(Result::unwrap_err).collect();

        let merge_errors = things
            .into_iter()
            .filter_map(|data| ArgVar::deep_merge(&mut map, data).err())
            .collect::<Vec<_>>();
        errors.extend(merge_errors);

        if !errors.is_empty() {
            return Err(errors);
        }

        Ok(mustache::Data::Map(map))
    }
}

impl ArgVar {
    fn deep_merge(
        map: &mut HashMap<String, mustache::Data>,
        data: HashMap<String, mustache::Data>,
    ) -> Result<(), ArgVarError> {
        ArgVar::deep_merge_impl(map, data, &Vec::new())
    }

    fn deep_merge_impl(
        map: &mut HashMap<String, mustache::Data>,
        data: HashMap<String, mustache::Data>,
        ctx: &[&str],
    ) -> Result<(), ArgVarError> {
        use std::collections::hash_map::Entry;

        for (key, value) in data {
            match map.entry(key) {
                Entry::Vacant(entry) => {
                    entry.insert(value);
                }
                Entry::Occupied(mut entry) => {
                    let key = entry.key().clone();
                    let entry_value = entry.get_mut();
                    match (entry_value, value) {
                        (mustache::Data::Map(inner_map), mustache::Data::Map(inner_value)) => {
                            let mut ctx = Vec::from(ctx);
                            ctx.push(&key);
                            ArgVar::deep_merge_impl(inner_map, inner_value, &ctx)?;
                        }
                        (mustache::Data::Map(inner_map), value) => {
                            dbg!(&ctx);
                            dbg!(&inner_map);
                            dbg!(&value);
                            return Err(ArgVarError::Merge);
                        }
                        (entry_value, mustache::Data::Map(inner_value)) => {
                            dbg!(&ctx);
                            dbg!(&entry_value);
                            dbg!(&inner_value);
                            return Err(ArgVarError::Merge);
                        }
                        (entry_value, value) => {
                            dbg!(&ctx);
                            dbg!(&entry_value);
                            dbg!(&value);
                            return Err(ArgVarError::Merge);
                        }
                    }
                }
            }
        }

        Ok(())
    }
}

#[derive(Debug, thiserror::Error)]
pub enum ArgVarError {
    #[error("{0:?}: empty key (variable must be in the form `key=value`)")]
    EmptyKey(String),
    #[error("{0:?}: empty value (variable must be in the form `key=value`)")]
    EmptyValue(String),
    #[error("{0:?}: too many `=` (variable must be in the form `key=value`)")]
    TooManyEquals(String),
    #[error("{0:?}: missing `=` (variable must be in the form `key=value`)")]
    NoEquals(String),
    #[error("missing (variable must be in the form `key=value`)")]
    NoContent,
    #[error("encountered errors when merging")]
    Merge,
}

#[allow(clippy::manual_try_fold)]
pub fn arg_vars_to_data<I>(iter: I, opt_name: &str) -> eyre::Result<mustache::Data>
where
    I: IntoIterator<Item = ArgVar>,
{
    use color_eyre::Section;
    use eyre::eyre;

    iter.into_iter().collect::<Result<_, _>>().or_else(|errs| {
        errs.into_iter().fold(
            Err(eyre!("encountered errors parsing --{opt_name}")),
            |report, e| report.error(e),
        )
    })
}
