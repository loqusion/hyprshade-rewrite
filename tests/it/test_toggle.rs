use proc_macros::hyprland_test;

use crate::common::{CommandExt, Space, hyprshade_cmd_snapshot};

#[test]
fn empty_arg_fails_without_config() {
    let mut space = Space::new();
    space.with_any_time();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle"), @r"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mno configuration file found[0m

    Location:
       [LOCATION]

    [93mWarning[0m: A configuration file is required to call this command without SHADER
    [96mSuggestion[0m: For more information, see [URL]
    ");
}

#[hyprland_test]
fn empty_arg_fails_resolving_default_shader() {
    let mut space = Space::new();
    space.with_any_time().with_config(
        r#"
            [[shader]]
            name = "invalid"
            default = true
        "#,
    );
    space.hyprshade_cmd().arg("off").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle"), @r#"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mresolving shader in config[0m
       1: [91mshader named "invalid" not found[0m

    Location:
       [LOCATION]

    Configuration:
       [33m[HYPRSHADE_CONFIG][39m

    [96mNote[0m: Since you omitted SHADER from cli arguments, it was inferred from the schedule in your configuration
    [96mSuggestion[0m: Change the shader name in your configuration, or make sure a shader by that name exists
    [96mSuggestion[0m: For more information, see [URL]
    "#);
}

#[hyprland_test]
fn empty_arg_fails_resolving_scheduled_shader() {
    let mut space = Space::new();
    space.with_time("01:00:00").with_config(
        r#"
            [[shader]]
            name = "invalid"
            start_time = 00:00:00
            end_time = 02:00:00
        "#,
    );

    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("on").arg("vibrance").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle"), @r#"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mresolving shader in config[0m
       1: [91mshader named "invalid" not found[0m

    Location:
       [LOCATION]

    Configuration:
       [33m[HYPRSHADE_CONFIG][39m

    [96mNote[0m: Since you omitted SHADER from cli arguments, it was inferred from the schedule in your configuration
    [96mSuggestion[0m: Change the shader name in your configuration, or make sure a shader by that name exists
    [96mSuggestion[0m: For more information, see [URL]
    "#);
}

#[hyprland_test]
fn empty_arg_with_active_shader_turns_off() {
    let mut space = Space::new();
    space.with_any_time().with_config(r#""#);

    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("on").arg("vibrance").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle"), @r"
    success: true
    exit_code: 0
    ----- stdout -----

    ----- stderr -----
    ");

    assert!(space.current_shader().is_none());
}

#[hyprland_test]
fn empty_arg_with_no_active_shader_turns_on_default() {
    let mut space = Space::new();
    space.with_any_time().with_config(
        r#"
            [[shader]]
            name = "vibrance"
            default = true
        "#,
    );

    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("off").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle"), @r"
    success: true
    exit_code: 0
    ----- stdout -----

    ----- stderr -----
    ");

    assert_eq!(space.current_shader().as_deref(), Some("vibrance"));
}

#[test]
fn fallback_default_fails_without_config() {
    let mut space = Space::new();
    space.with_any_time();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-default"]), @r"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mno configuration file found[0m

    Location:
       [LOCATION]

    [93mWarning[0m: A configuration file is required to use --fallback-default
    [96mSuggestion[0m: For more information, see [URL]
    ");
}

#[test]
fn fallback_default_fails_without_default_shader() {
    let mut space = Space::new();
    space.with_any_time().with_config(r#""#);

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-default"]), @r"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mno default shader found in config[0m

    Location:
       [LOCATION]

    Configuration:
       [33m[HYPRSHADE_CONFIG][39m

    [96mSuggestion[0m: Make sure a default shader is defined (default = true)
    [96mSuggestion[0m: For more information, see [URL]
    ");
}

#[hyprland_test]
fn fallback_default_fails_resolving_shader() {
    let mut space = Space::new();
    space.with_any_time().with_config(
        r#"
            [[shader]]
            name = "invalid"
            default = true
        "#,
    );

    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("on").arg("vibrance").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-default"]), @r#"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mresolving default shader in config[0m
       1: [91mshader named "invalid" not found[0m

    Location:
       [LOCATION]

    Configuration:
       [33m[HYPRSHADE_CONFIG][39m

    [96mSuggestion[0m: Change the shader name in your configuration, or make sure a shader by that name exists
    [96mSuggestion[0m: For more information, see [URL]
    "#);
}

#[hyprland_test]
fn fallback_default_with_active_shader_turns_on_default() {
    let mut space = Space::new();
    space.with_any_time().with_config(
        r#"
            [[shader]]
            name = "color-filter"
            default = true
        "#,
    );

    let _stash = space.stash_runtime_shaders(["color-filter", "vibrance"]);
    space.hyprshade_cmd().arg("on").arg("vibrance").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-default"]), @r"
    success: true
    exit_code: 0
    ----- stdout -----

    ----- stderr -----
    ");

    assert_eq!(space.current_shader().as_deref(), Some("color-filter"));
}

#[hyprland_test]
fn fallback_default_with_default_turns_on_positional_arg() {
    let mut space = Space::new();
    space.with_any_time().with_config(
        r#"
            [[shader]]
            name = "color-filter"
            default = true
        "#,
    );

    let _stash = space.stash_runtime_shaders(["color-filter", "vibrance"]);
    space.hyprshade_cmd().arg("on").arg("color-filter").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-default"]), @r"
    success: true
    exit_code: 0
    ----- stdout -----

    ----- stderr -----
    ");

    assert_eq!(space.current_shader().as_deref(), Some("vibrance"));
}

#[test]
fn fallback_auto_fails_without_config() {
    let mut space = Space::new();
    space.with_any_time();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-auto"]), @r"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mno configuration file found[0m

    Location:
       [LOCATION]

    [93mWarning[0m: A configuration file is required to use --fallback-auto
    [96mSuggestion[0m: For more information, see [URL]
    ");
}

#[hyprland_test]
fn fallback_auto_fails_resolving_scheduled_shader() {
    let mut space = Space::new();
    space.with_time("01:00:00").with_config(
        r#"
            [[shader]]
            name = "invalid"
            start_time = 00:00:00
            end_time = 02:00:00
        "#,
    );

    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("on").arg("vibrance").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-auto"]), @r#"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mresolving shader in config[0m
       1: [91mshader named "invalid" not found[0m

    Location:
       [LOCATION]

    Configuration:
       [33m[HYPRSHADE_CONFIG][39m

    [96mNote[0m: Tried to resolve scheduled shader because of --fallback-auto
    [96mSuggestion[0m: Change the shader name in your configuration, or make sure a shader by that name exists
    [96mSuggestion[0m: For more information, see [URL]
    "#);
}

#[hyprland_test]
fn fallback_auto_fails_resolving_default_shader() {
    let mut space = Space::new();
    space.with_time("01:00:00").with_config(
        r#"
            [[shader]]
            name = "invalid"
            default = true
            [[shader]]
            name = "vibrance"
            start_time = 00:00:00
            end_time = 02:00:00
        "#,
    );

    let _stash = space.stash_runtime_shader("vibrance");
    space.hyprshade_cmd().arg("on").arg("vibrance").run();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args(["vibrance", "--fallback-auto"]), @r#"
    success: false
    exit_code: 1
    ----- stdout -----

    ----- stderr -----
    Error: 
       0: [91mresolving default shader in config[0m
       1: [91mshader named "invalid" not found[0m

    Location:
       [LOCATION]

    Configuration:
       [33m[HYPRSHADE_CONFIG][39m

    [96mNote[0m: Tried to resolve default shader because of --fallback-auto
    [96mSuggestion[0m: Change the shader name in your configuration, or make sure a shader by that name exists
    [96mSuggestion[0m: For more information, see [URL]
    "#);
}

#[hyprland_test]
fn fallback_auto_success() {
    #[derive(Clone, Copy, Debug)]
    enum TestVariant {
        WithDefault,
        NoDefault,
    }

    #[derive(Clone, Debug)]
    struct Case {
        time: &'static str,
        initial_shader: &'static str,
        expected_shader: ExpectedShader,
    }
    #[derive(Clone, Debug)]
    struct ExpectedShader {
        with_default: Option<&'static str>,
        no_default: Option<&'static str>,
    }
    impl Case {
        fn snapshot_name(&self, variant: TestVariant) -> String {
            let suffix = match &variant {
                TestVariant::NoDefault => "no-default",
                TestVariant::WithDefault => "with-default",
            };
            format!(
                "fallback_auto_success-{}-{}-{}-{}-{}",
                self.time,
                self.initial_shader,
                self.expected_shader.with_default.unwrap_or("--NONE--"),
                self.expected_shader.no_default.unwrap_or("--NONE--"),
                suffix
            )
        }
    }

    const STASHED_SHADERS: &[&str] = &[
        "blue-light-filter",
        "color-filter",
        "invert-colors",
        "vibrance",
    ];
    const CONFIG_WITH_DEFAULT: &str = r#"
        [[shader]]
        name = "color-filter"
        default = true

        [[shader]]
        name = "blue-light-filter"
        start_time = 01:00:00
        end_time = 03:00:00

        [[shader]]
        name = "invert-colors"
        start_time = 23:00:00
        end_time = 01:00:00
    "#;
    const CONFIG_NO_DEFAULT: &str = r#"
        [[shader]]
        name = "blue-light-filter"
        start_time = 01:00:00
        end_time = 03:00:00

        [[shader]]
        name = "invert-colors"
        start_time = 23:00:00
        end_time = 01:00:00
    "#;

    #[track_caller]
    fn check(case: Case) {
        #[track_caller]
        fn _check(case: &Case, variant: TestVariant) {
            let (config, expected_shader) = match variant {
                TestVariant::WithDefault => {
                    (CONFIG_WITH_DEFAULT, &case.expected_shader.with_default)
                }
                TestVariant::NoDefault => (CONFIG_NO_DEFAULT, &case.expected_shader.no_default),
            };

            let Case {
                time,
                initial_shader,
                expected_shader: _,
            } = &case;

            let mut space = Space::new();
            let _stash = space.stash_runtime_shaders(STASHED_SHADERS);
            space.with_time(time).with_config(config);

            space.hyprshade_cmd().arg("on").arg(initial_shader).run();

            hyprshade_cmd_snapshot!(
                case.snapshot_name(variant),
                space
                    .hyprshade_cmd()
                    .arg("toggle")
                    .args(["vibrance", "--fallback-auto"])
            );

            assert_eq!(
                space.current_shader().as_deref(),
                *expected_shader,
                "{:?}",
                variant
            );
        }

        _check(&case, TestVariant::WithDefault);
        _check(&case, TestVariant::NoDefault);
    }

    check(Case {
        time: "00:00:00",
        initial_shader: "vibrance",
        expected_shader: ExpectedShader {
            with_default: Some("invert-colors"),
            no_default: Some("invert-colors"),
        },
    });
    check(Case {
        time: "00:00:00",
        initial_shader: "blue-light-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "00:00:00",
        initial_shader: "invert-colors",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "00:00:00",
        initial_shader: "color-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });

    check(Case {
        time: "00:59:59.999999999",
        initial_shader: "vibrance",
        expected_shader: ExpectedShader {
            with_default: Some("invert-colors"),
            no_default: Some("invert-colors"),
        },
    });
    check(Case {
        time: "00:59:59.999999999",
        initial_shader: "blue-light-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "00:59:59.999999999",
        initial_shader: "invert-colors",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "00:59:59.999999999",
        initial_shader: "color-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });

    check(Case {
        time: "01:00:00",
        initial_shader: "vibrance",
        expected_shader: ExpectedShader {
            with_default: Some("blue-light-filter"),
            no_default: Some("blue-light-filter"),
        },
    });
    check(Case {
        time: "01:00:00",
        initial_shader: "blue-light-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "01:00:00",
        initial_shader: "invert-colors",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "01:00:00",
        initial_shader: "color-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });

    check(Case {
        time: "02:00:00",
        initial_shader: "vibrance",
        expected_shader: ExpectedShader {
            with_default: Some("blue-light-filter"),
            no_default: Some("blue-light-filter"),
        },
    });
    check(Case {
        time: "02:00:00",
        initial_shader: "blue-light-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "02:00:00",
        initial_shader: "invert-colors",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "02:00:00",
        initial_shader: "color-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });

    check(Case {
        time: "02:59:59.999999999",
        initial_shader: "vibrance",
        expected_shader: ExpectedShader {
            with_default: Some("blue-light-filter"),
            no_default: Some("blue-light-filter"),
        },
    });
    check(Case {
        time: "02:59:59.999999999",
        initial_shader: "blue-light-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "02:59:59.999999999",
        initial_shader: "invert-colors",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "02:59:59.999999999",
        initial_shader: "color-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });

    check(Case {
        time: "03:00:00",
        initial_shader: "vibrance",
        expected_shader: ExpectedShader {
            with_default: Some("color-filter"),
            no_default: None,
        },
    });
    check(Case {
        time: "03:00:00",
        initial_shader: "blue-light-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "03:00:00",
        initial_shader: "invert-colors",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
    check(Case {
        time: "03:00:00",
        initial_shader: "color-filter",
        expected_shader: ExpectedShader {
            with_default: Some("vibrance"),
            no_default: Some("vibrance"),
        },
    });
}

#[test]
fn fallback_flags_are_mutually_exclusive() {
    let space = Space::new();

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args([
        "--fallback",
        "vibrance",
        "--fallback-auto",
        "blue-light-filter",
    ]), @r"
    success: false
    exit_code: 2
    ----- stdout -----

    ----- stderr -----
    error: the argument '--fallback <FALLBACK>' cannot be used with '--fallback-auto'

    Usage: hyprshade toggle --fallback <FALLBACK> <SHADER>

    For more information, try '--help'.
    ");

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args([
        "--fallback",
        "vibrance",
        "--fallback-default",
        "blue-light-filter",
    ]), @r"
    success: false
    exit_code: 2
    ----- stdout -----

    ----- stderr -----
    error: the argument '--fallback <FALLBACK>' cannot be used with '--fallback-default'

    Usage: hyprshade toggle --fallback <FALLBACK> <SHADER>

    For more information, try '--help'.
    ");

    hyprshade_cmd_snapshot!(space.hyprshade_cmd().arg("toggle").args([
        "--fallback-auto",
        "--fallback-default",
        "blue-light-filter",
    ]), @r"
    success: false
    exit_code: 2
    ----- stdout -----

    ----- stderr -----
    error: the argument '--fallback-auto' cannot be used with '--fallback-default'

    Usage: hyprshade toggle --fallback-auto <SHADER>

    For more information, try '--help'.
    ");
}

#[test]
#[ignore = "already tested in test_on::error::fails_merging_var"]
fn fails_merging_var() {
    unimplemented!()
}

#[test]
#[ignore = "should be near-identical to fails_merging_var"]
fn fails_merging_fallback_var() {
    unimplemented!()
}

#[test]
#[ignore = "already tested in test_on::error::fails_parsing_var"]
fn fails_parsing_var() {
    unimplemented!()
}

#[test]
#[ignore = "should be near-identical to fails_parsing_var"]
fn fails_parsing_fallback_var() {
    unimplemented!()
}
