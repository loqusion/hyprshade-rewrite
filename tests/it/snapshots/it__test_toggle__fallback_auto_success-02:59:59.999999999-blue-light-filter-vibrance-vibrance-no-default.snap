---
source: tests/it/test_toggle.rs
info:
  program: hyprshade
  args:
    - toggle
    - vibrance
    - "--fallback-auto"
  env:
    COLORBT_SHOW_HIDDEN: ""
    HOME: /tmp/.tmpAQnJ0A/hyprshade-test-dir/home/
    RUST_BACKTRACE: ""
    XDG_CONFIG_HOME: /tmp/.tmpAQnJ0A/hyprshade-test-dir/home/.config
    __HYPRSHADE_MOCK_TIME: "02:59:59.999999999"
---
success: true
exit_code: 0
----- stdout -----

----- stderr -----
